In [5]:
import numpy as np, pandas as pd, matplotlib.pyplot as plt, mlflow
from sklearn.datasets import fetch_california_housing
from sklearn.base import clone
from sklearn.model_selection import train_test_split, cross_val_score, KFold
from sklearn.linear_model import LinearRegression, LogisticRegression, Ridge, Lasso
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline


In [6]:
data = fetch_california_housing(as_frame=True)
X = data.data
y = data.target
pipe = make_pipeline(StandardScaler(), LinearRegression())

mlflow.set_experiment(experiment_name='week-02-ridge-lasso')

with mlflow.start_run(run_name='linreg-5fold-cv'):
    scores = cross_val_score(pipe, X, y, cv=5, scoring='r2')
    mlflow.log_param('cv_folds', '5')
    mlflow.log_param('model_type', 'linreg')
    print(scores)
    print(scores.mean())
    print(scores.std())
    mlflow.log_metric('cv_mean_r2', scores.mean())
    mlflow.log_metric('cv_std_r2', scores.std())
    f"{scores.mean():.4f} ± {scores.std():.4f}"

[0.54866323 0.46820691 0.55078434 0.53698703 0.66051406]
0.5530311140279563
0.061691601409532774


From single split the r2score is : 0.5757, and from cv, mean r2score is 0.5530 
single split is over estimated r2score by 0.0227
std is 0.0617 which means r2score could range from 0.4913 (0.5530-0.0617) to 0.6147 (0.5530+0.0617)

In [7]:
with mlflow.start_run(run_name='ridge_5fold_cv'):
    pipe = make_pipeline(StandardScaler(), Ridge(1.0))

    mlflow.log_param('cv_folds', '5')
    mlflow.log_param('alpha_used', '1.0')
    mlflow.log_param('model_type', 'ridge_reg')

    scores = cross_val_score(pipe, X, y, cv=5, scoring='r2' )

    print(scores)
    print(scores.mean())
    print(scores.std())
    mlflow.log_metric('cv_mean_r2', scores.mean())
    mlflow.log_metric('cv_std_r2', scores.std())
    f"{scores.mean():.4f} ± {scores.std():.4f}"

[0.54874936 0.46818182 0.55080227 0.53691869 0.66051907]
0.55303424082049
0.0617024714315415


For Ridge regression, 
From the single split train test dataset, the r2score is 0.5758 and from cv r2score is 0.5530 
single split is over predicted by 0.228, which very similar to linear regression
std deviation also same as linear regression,
it tells you that ridge regression barely changes anything from linear regression here

In [8]:
with mlflow.start_run(run_name='lasso-5fold-cv'):
    pipe = make_pipeline(StandardScaler(), Lasso(0.1))
    mlflow.log_param('cv_folds', '5')
    mlflow.log_param('alpha_used', '0.1')
    mlflow.log_param('model_type', 'lasso_reg')
    scores = cross_val_score(pipe, X, y, cv=5, scoring='r2' )
    print(scores)
    print(scores.mean())
    print(scores.std())
    mlflow.log_metric('cv_mean_r2', scores.mean())
    mlflow.log_metric('cv_std_r2', scores.std())
    print(f"{scores.mean():.4f} ± {scores.std():.4f}")

[0.45509221 0.40234689 0.47091764 0.30468546 0.52259382]
0.43112720517469005
0.0739453362418414
0.4311 ± 0.0739


For Lasso regression, 
from the single split train-test dataset, the value is 0.4813 and from cv is 0.4311 
single split is over predicted by 0.0502
std deviation is 0.0739 which means that r2score can range from 0.3572 to 0.505,
So by considering the above models Linear regression and Ridge regression are simillar so any one can do better, and both models has simillar mean 0.5530 and as std deviation also same and can range from 0.4913 (0.5530-0.0617) to 0.6147 (0.5530+0.0617)